docs/samediff/templates/variables.md

```diff
@@ -1,3 +1,214 @@
+
+To infer the data type of your variable, use
+```java
+DataType dataType = yourVariable.dataType();
+```
+You may need to trace your variable's data type since at times it does matter, which types you use in an operation. For 
+example, a convolution product, like this one
+```java
+SDVariable prod = samediff.cnn.conv1d(input, weights, config);
+```
+will require its `SDVariable` arguments `input` and `weights` to be of one of the floating point data types, and will
+throw an exception otherwise. Also, as we shall discuss just below, all the `SDVariables` of type `VARIABLE` are 
+supposed to be of floating point type.
+
+## Common features of variables
+Before we go to the differences between variables, let us first look at the properties they all share 
+- All variables are ultimately derived from an instance of `SameDiff`, serving as parts of its 
+[graph](./samediff/graphs). In fact, each variable has a `SameDiff` as one of its fields.
+- Results (outputs) of all operations are of `ARRAY` type. 
+- All `SDVariable`'s involved in an operation are to belong to the *same* `SameDiff`. 
+- All variables may or may not be given names - in the latter case, a name is actually created automatically. Either
+way, the names need to be/are created unique. We shall come back to naming below.
+
+## Differences between variable types
+
+Let us now have a closer look at each type of variables, and what distinguish them from each other.
+
+### Variables
+
+Variables are the trainable parameters of your network. This predetermines their nature in `SameDiff`. As we briefly 
+mentioned above, variables' values need to be 
+both preserved for application, and updated during training. Training means, that we iteratively 
+update the values by small fractions of their gradients, and this only makes sense if variables are of *floating 
+point* types (see data types above).
+
+Variables may be added to your `SameDiff` using different versions of `var` function from your `SameDiff` instance. 
+For example, the code
+```java
+SDVariable weights = samediff.var("weights", DataType.FLOAT, 784, 10);
+```
+adds a variable constituting of a 784x10 array of `float` numbers - weights for a single layer MNIST perceptron 
+in this case - to a pre-existing `SameDiff` instance `samediff`.
+However, this way the values within a variable will be set as zeros. If we wish them to be filled, say, with random 
+numbers, we need to add a weight initialization scheme, for instance like that:
+```java
+SDVariable weights = samediff.var("weights", new XavierInitScheme('c', 784, 10), DataType.FLOAT, 784, 10);
+```
+Now, the weights will be randomly initialized using the Xavier scheme. There are other ways to create and 
+fill variables, you may see a full reference in our [javadoc](https://deeplearning4j.org/api/latest/).
```

Comment:
Can we link more directly?
"see the 'known subclasses' section of https://deeplearning4j.org/api/latest/org/nd4j/weightinit/WeightInitScheme.html"

Also add a sd.var(..., INDArray) example, with a manually (Nd4j.someMethod) created array.